### PR TITLE
feat: allow sending messages to stdin

### PR DESCRIPTION
**Pull request by wash2**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -160,7 +187,7 @@ impl ProcessManager {
 			.envs(process_data.process.env.clone())
 			.stdout(Stdio::piped())
 			.stderr(Stdio::piped())
-			.stdin(Stdio::null())
+			.stdin(Stdio::piped())
```
Comment (ids1024) [inline src/lib.rs]: So, this always uses a pipe for stdin?  Will uses of this always/usually want a pipe?

I guess it isn't too much of a problem, though,  to have a pipe that nothing writes to. Insignificantly more overhead. (Unlike an stdout/stderr pipe, which needs to be read or the process will block when the pipe buffer is filled.)

Comment (wash2) [reply src/lib.rs]: Most uses of this probably won't usually want a pipe. I can refactor this to make it optional?

**Comment by wash2** [comment]
Also, this includes callbacks for running immediately before and after a process is started. These callbacks are currently used in cosmic-sessions and cosmic-panel for setting and un-setting cloexec. I'm just getting around to sorta testing it all though.

File: src/lib.rs (hunk 1 above)
Comment (ids1024) [reply src/lib.rs]: In principle I think it shouldn't be creating a pipe when it normally isn't needed, but that may be simpler in some ways.

**Comment by ids1024** [comment]
Hm. Unsetting/setting `CLOEXEC` that way is risky. In a multi-threaded environment. (If nothing else, cosmic-panel has Mesa background threads.) Fd flags like this apply to all threads, so if another thread forks and execs a new child process, while the fd isn't marked `CLOEXEC`, that other process will also get the fd. Which may be a security concern and otherwise lead to unexpected behavior.

`open(2)` notes that `O_CLOEXEC` is needed in multi-threaded programs, instead of just setting `FD_CLOEXEC` after opening, for exactly this reason.

`CLOEXEC` is  not the most elegant part of POSIX.

If necessary then I believe the correct thing to do is to unset cloexec after forking but before exec, which https://doc.rust-lang.org/std/os/unix/process/trait.CommandExt.html#tymethod.pre_exec can do. Though that API is rather unsafe because many things are unsafe to do after forking in a multi-threaded process.

**Comment by wash2** [comment]
> Hm. Unsetting/setting `CLOEXEC` that way is risky. In a multi-threaded environment. (If nothing else, cosmic-panel has Mesa background threads.) Fd flags like this apply to all threads, so if another thread forks and execs a new child process, while the fd isn't marked `CLOEXEC`, that other process will also get the fd. Which may be a security concern and otherwise lead to unexpected behavior.

Ya, I guess I assumed that since the panel and session are only launching new processes in the launch-pad thread it might be alright. I'll look into https://doc.rust-lang.org/std/os/unix/process/trait.CommandExt.html#tymethod.pre_exec though!

**Comment by ids1024** [comment]
Yeah, there's *probably* not anything that would be forking in the background, but it's hard to know exactly what libc, mesa, or any other libraries might be doing.

At least there should be a comment in cosmic-panel/cosmic-session noting this, and why it should be safe. But it would be good to do it the correct way if possible.

**Comment by ids1024** [comment]
Suppose for instance that libc spawned Perl to deal with strings because C is too hard. But no serious OS would do that. :laughing:
https://news.ycombinator.com/item?id=18920122

(The original tweet won't load without being logged in, apparently, because of course Twitter changed that today.)

File: src/lib.rs [hunk 2]
```diff
@@ -176,25 +230,55 @@ impl ProcessManager {
...
-			.map_err(Error::Process)?;
+		let command = unsafe {
+			Command::new(&process_data.process.executable)
+				.args(&process_data.process.args)
+				.envs(process_data.process.env.clone())
+				.stdout(Stdio::piped())
+				.stderr(Stdio::piped())
+				.stdin(Stdio::piped())
+				.kill_on_drop(true)
+				.pre_exec({
```
Comment (ids1024) [inline src/lib.rs]: Seems a bit odd to use a block here. I was confused before I realized the first part wasn't part of the closure. Might be clearer to move `let fd_list = ...` to before `let command`?

But more importantly, this implementation looks like it should be correct and sound. Just calling `mark_as_not_cloexec()` should hopefully be sound, while doing much more in `pre_exec` may not be.  :+1:

File: src/lib.rs [hunk 3]
```diff
@@ -112,42 +113,90 @@ impl ProcessManager {
...
+						Ok(())
+					}
+				})
+				.spawn()
+				.map_err(Error::Process)?
+		};
+		if let Some(fds) = process.process.callbacks.fds.as_ref() {
+			let fd_list = fds();
+			for fd in fd_list {
+				if let Err(err) = util::mark_as_cloexec(fd) {
```
Comment (ids1024) [inline src/lib.rs]: If we call `mark_as_not_cloexec` only in `pre_exec`, that will only apply in the spawned subprocess, so we don't needed to mark as cloexec here.

And the `mark_as_cloexec` helper should just be unnecessary now.

**Review by ids1024 — APPROVED** [review]

**Comment by ids1024** [comment]
Something to be aware of using this is that the spawned process should immediately set `CLOEXEC` (hopefully before spawning any threads) to avoid running into the same issue. With libwayland/wayland-rs that will happen automatically with `WAYLAND_SOCKET` is used. But if this is used for anything else, that's something to make sure is handled.

**Comment by wash2** [comment]
> Something to be aware of using this is that the spawned process should immediately set CLOEXEC (hopefully before spawning any threads) to avoid running into the same issue. With libwayland/wayland-rs that will happen automatically with WAYLAND_SOCKET is used. But if this is used for anything else, that's something to make sure is handled.

Ok thanks, I'll make sure to do so in the notifications-daemon and panel

**Pull request merged**